components/eam/src/physics/cam/constituents.F90

```diff
@@ -435,4 +439,18 @@ end function cnst_cam_outfld
 
 !==============================================================================
 
+
+subroutine setup_moist_indices()
+
+   implicit none
+
+   call cnst_get_ind('CLDICE', icldice, abrtf=.false.)
+   call cnst_get_ind('CLDLIQ', icldliq, abrtf=.false.)
+   call cnst_get_ind('RAINQM', irain, abrtf=.false.)
+   call cnst_get_ind('SNOWQM', isnow, abrtf=.false.)
+
+ end subroutine setup_moist_indices
```

Comment:
I don't think it's a good idea to do this. The constituent interface is meant to be generic so that the physics package can decide what constituents need to be added. It's pretty clear that the error you noticed would be from not calling setup_moist_indices() in the MMF physpkg.

The constituent names for the MMF are set in crm_physics.F90. We could move this to the MMF specific physpkg, but I thought it made since to have in crm_physics so it could be modified according to the CRM options (i.e. microphysics and variance tracers). 

Currently, check_energy_register() is called before crm_physics_register() in the MMF physpkg, but there's no reason we couldn't change the order of this so that check_energy_register() can set the indices correctly.

So I think a better strategy would be to have local indices in check_energy.F90 (and wherever else they are needed) and have the values set in check_energy_register() after the call to crm_physics_register().

Reply:
I agree with Walter. I think constituents.F90 provides all the get/set functions for individual species, which keeps it generic.